File: lib/gherkin/parser.rb
module Gherkin
  class ParseError < StandardError
    def initialize(state, new_state, expected_states, line)
      super("Parse error on line #{line}. Found #{new_state} when expecting one of: #{expected_states.join(', ')}. (Current state: #{state}).")
    end
  end

  class Parser
    def self.new(listener, raise_on_error=false, machine_name='root')
      if defined?(JRUBY_VERSION)
        require 'gherkin.jar'
        Java::Gherkin::Parser.new(listener, raise_on_error, machine_name)
      else
        require 'gherkin/rb_parser'
        Gherkin::RbParser.new(listener, raise_on_error, machine_name)
      end
    end
  end
end

File: dotnet/Gherkin.Tests/ParserSpecs/Parser.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Linq.Expressions;
using Moq;
using Xunit;
using It=Moq.It;

namespace Gherkin.Tests.ParserSpecs
{
    public class Parser : ParserSpec
    {
        [Fact]
        public void should_delegate_events_to_the_listener()
        {
            a_parser().
                parsing_input("# Content").
                should_delegate(l => l.Comment(new Token("# Content", new Position(1, 1))));
        }

        [Fact]
        public void should_raise_helpful_error_messages_by_default()
        {
            a_parser().
                receiving(p => p.Scenario(
                    new Token("Scenario", new Position(12,1)), 
                    new Token("My pet scenario", new Position(12, 10)))).
                should_raise_error("Parse error on line 12. Found scenario when expecting one of: comment, eof, feature, tag. (Current state: root).");
        }

        [Fact]
        public void should_delegate_an_error_message_when_raise_on_error_is_false()
        {
            a_parser()
                .with_raise_on_error(false)
                .receiving(p => p.Background(
                    new Token("Background", new Position(1, 1)), 
                    new Token("Content", new Position(1, 12))))
                .should_delegate(
                l => l.SyntaxError("root", "background", Containing<string>.Same("comment", "eof", "feature", "tag"), It.Is<Position>(p => p.Line == 1)));
        }
    }

    public static class Containing<TItem>
    {
        public static IEnumerable<TItem> Same(params TItem[] values)
        {
            return
                Match<IEnumerable<TItem>>.Create(
                    input => input.OrderBy(item => item).SequenceEqual(values.OrderBy(item => item)));
        }
    }
}
